chainer/functions/normalization/batch_normalization.py

```diff
@@ -213,16 +213,18 @@ def forward(self, inputs):
         else:
             # Generic CPU and GPU implementation
 
-            gamma = gamma[expander]
-            beta = beta[expander]
-            self.mean = x.mean(axis=self.axis, dtype=gamma.dtype)
-            var = x.var(axis=self.axis, dtype=gamma.dtype)
+            dtype = numpy.result_type(x, gamma)
+
+            gamma = gamma[expander].astype(dtype, copy=False)
+            beta = beta[expander].astype(dtype, copy=False)
+            self.mean = x.mean(axis=self.axis, dtype=dtype)
+            var = x.var(axis=self.axis, dtype=dtype)
             if xp is numpy:
                 self.inv_std = numpy.reciprocal(numpy.sqrt(
-                    var + self.eps, dtype=gamma.dtype))
+                    var + self.eps, dtype=dtype))
```

Comment:
Passing `gamma.dtype` assumes that `gamma` has a higher precision than `x`, which is (true for `mixed16` but) not always the case.